libsolidity/ast/Types.cpp

```diff
@@ -353,16 +508,57 @@ bool IntegerConstantType::isValidLiteral(const Literal& _literal)
 	return true;
 }
 
-IntegerConstantType::IntegerConstantType(Literal const& _literal)
+ConstantNumberType::ConstantNumberType(Literal const& _literal)
 {
-	m_value = bigint(_literal.value());
-
+	rational numerator;
+	rational denominator = bigint(1);
+	m_scalingFactor = 0;
+	auto radixPoint = find(_literal.value().begin(), _literal.value().end(), '.');
+	if (radixPoint != _literal.value().end())
+	{
+		auto leadingZeroes = find_if_not(
+				radixPoint + 1, 
+				_literal.value().end(), 
+				[](char const& a) { return a == '0'; }
+			);
+		auto fractionalBegin = leadingZeroes != _literal.value().end() ?
+			leadingZeroes : radixPoint + 1;
+		//separatly grabbing the numerator, denominator, and the scaling factor for conversions
+		string fractionals = string(fractionalBegin, _literal.value().end());
+		m_scalingFactor = distance(radixPoint + 1, _literal.value().end());
+		denominator = bigint(fractionals);
+		denominator /= boost::multiprecision::pow(bigint(10), m_scalingFactor);
+		numerator = bigint(string(_literal.value().begin(), radixPoint));
+
+		m_value = numerator + denominator;		
+		//cout << "Original form: " << _literal.value() << endl;
+		//cout << "Actual form: " << to_string(m_value) << endl;
+		//cout << "Float form: " << to_string(boost::rational_cast<long double>(m_value)) << endl;
+		//cout << endl;
+	}
+	else
+	{
+		numerator = bigint(_literal.value());
+		m_value.assign(numerator.numerator(), denominator.numerator());
+	}
 	switch (_literal.subDenomination())
 	{
+	case Literal::SubDenomination::None:
+		break;
 	case Literal::SubDenomination::Wei:
 	case Literal::SubDenomination::Second:
-	case Literal::SubDenomination::None:
+	{
+		if (m_value.denominator() != 1)
+		{
+			char const* sub = Token::toString(_literal.token());
+			BOOST_THROW_EXCEPTION(
+				Error(Error::Type::TypeError) << errinfo_comment(
+					"Cannot divide sub denomination " + ((sub) ? string(Token::toString(_literal.token())) : "") + " any further."
```

Comment:
Why? `ufixed x = .5 wei` is perfectly fine, you will get an error when you want to convert it to `uint`.